src/generators/component/templates/react/component-with-d3.js

```diff
@@ -1,77 +1,47 @@
-import React from 'react';
+import React, { useRef, useLayoutEffect, useEffect } from 'react';
 import { select } from 'd3-selection';
-
 import styles from './styles.scss';
 
-export default class <%= className %> extends React.Component {
-  constructor(props) {
-    super(props);
-
-    this.rootRef = React.createRef();
-
-    this.initGraph = this.initGraph.bind(this);
-    this.updateGraph = this.updateGraph.bind(this);
-  }
-
-  componentWillReceiveProps(nextProps) {
-    // TODO: Add any conditions that mitigate updating the graph
-    this.updateGraph(nextProps);
-  }
-
-  shouldComponentUpdate() {
-    // Stop React from managing the DOM itself
-    return false;
-  }
-
-  componentDidMount() {
-    this.initGraph(this.props);
-
-    // TODO: add any listeners here
-    // ...
-  }
-
-  componentWillUnmount() {
-    // TODO: remove any listeners here
-    // ...
-  }
+let svg;
+let group;
+let rect;
 
-  /**
-   * Initialize the graph
-   * @param {object} props The latest props that were given to this component
-   */
-  initGraph(props) {
-    if (!this.rootRef.current) {
-      return;
-    }
+export default props => {
+  const root = useRef();
 
-    this.svg = select(this.rootRef.current)
+  useLayoutEffect(() => {
+    // Initialise the SVG
+    svg = select(root.current)
       .append('svg')
       .attr('width', 400)
       .attr('height', 300);
 
-    this.g = this.svg.append('g').attr('fill', 'black');
+    group = svg.append('g').attr('fill', 'black');
 
-    this.rect = this.g
+    rect = group
       .append('rect')
       .attr('x', 0)
       .attr('y', 0)
       .attr('rx', 3)
       .attr('ry', 3)
       .attr('width', 400)
       .attr('height', 300);
-  }
+  }, []);
 
-  /**
-   * Update the graph. It is important to only update this component through normal D3 methods.
-   * @param {object} props The latest props given to this component
-   */
-  updateGraph(props) {
-    if (!this.rootRef.current) return;
+  useEffect(() => {
+    updateChart(props);
+  }, [props]); // TODO: be more specific with your change checkers
 
-    // TODO: Use D3 to update the graph
-  }
+  // Example effect to update chart on window resize
+  useEffect(() => {
+    const onResize = () => updateChart(props);
+    window.addEventListener('resize', onResize);
+    return () => window.removeEventListener('resize', onResize);
+  }, []);
```

Comment:
I think this is still going to cause trouble. Subsequent invocations of the onResize function will pass updateChart the old props.

Reply:
I've added `props` as a dependency to the effect.